File: Data Structures & Algorithms/find-duplicate-integer/submission-0.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        hmap = {None: None}
        n = len(nums)
        i = 0
        
        while n:
            if nums[i] not in hmap:
                hmap[nums[i]] = 1
            else:
                return nums[i]
            n -= 1
            i += 1
        

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> postFix(nums.size());
        vector<int> preFix(nums.size());
        int index = 1;
        for(int i = 0; i < nums.size(); i++) {
            index *= nums[i];
            preFix[i] = index;
        }
        index = 1;
        for(int i = nums.size()-1; i > 0; i--){
            index *=nums[i];
            postFix[i] = index;
        }

        for(int i = 0; i < nums.size(); i++){
            if(i == 0) nums[i] = postFix[i+1];
            else if(i == nums.size()-1) nums[i] = preFix[i-1];
            else nums[i] = postFix[i+1] * preFix[i-1];
        }

        return nums;
    }
};
